### PR TITLE
step 0 of cuDNN v8 convolution API integration

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR is step 0 of adding PyTorch convolution bindings using the cuDNN frontend. The cuDNN frontend is the recommended way of using cuDNN v8 API. It is supposed to have faster release cycles, so that, for example, if people find a specific kernel has a bug, they can report it, and that kernel will be blocked in the cuDNN frontend and frameworks could just update that submodule without the need for waiting for a whole cuDNN release.

The work is not complete, and this PR is only step 0.

**What this PR does:**
- Add cudnn-frontend as a submodule.
- Modify cmake to build that submodule.
- Add bindings for convolution forward in `Conv_v8.cpp`, which is disabled by a macro by default.
- Tested manually by enabling the macro and run `test_nn.py`. All tests pass except those mentioned below.

**What this PR doesn't:**
- Only convolution forward, no backward. The backward will use v7 API.
- No 64bit-indexing support for some configuration. This is a known issue of cuDNN, and will be fixed in a later cuDNN version. PyTorch will not implement any workaround for issue, but instead, v8 API should be disabled on problematic cuDNN versions.
- No test beyond PyTorch's unit tests.
  - Not tested for correctness on real models.
  - Not benchmarked for performance.
- Benchmark cache is not thread-safe. (This is marked as `FIXME` in the code, and will be fixed in a follow-up PR)
- cuDNN benchmark is not supported.
- There are failing tests, which will be resolved later:
  ```
  FAILED test/test_nn.py::TestNNDeviceTypeCUDA::test_conv_cudnn_nhwc_cuda_float16 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=0.001 and atol=1e-05, found 32 element(s) (out of 32) whose difference(s) exceeded the margin of error (in...
  FAILED test/test_nn.py::TestNNDeviceTypeCUDA::test_conv_cudnn_nhwc_cuda_float32 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1.3e-06 and atol=1e-05, found 32 element(s) (out of 32) whose difference(s) exceeded the margin of error (...
  FAILED test/test_nn.py::TestNNDeviceTypeCUDA::test_conv_large_cuda - RuntimeError: CUDNN_BACKEND_OPERATION: cudnnFinalize Failed cudnn_status: 9
  FAILED test/test_nn.py::TestNN::test_Conv2d_depthwise_naive_groups_cuda - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=0 and atol=1e-05, found 64 element(s) (out of 64) whose difference(s) exceeded the margin of error (including 0 an...
  FAILED test/test_nn.py::TestNN::test_Conv2d_deterministic_cudnn - RuntimeError: not supported yet
  FAILED test/test_nn.py::TestNN::test_ConvTranspose2d_groups_cuda_fp32 - RuntimeError: cuDNN error: CUDNN_STATUS_BAD_PARAM
  FAILED test/test_nn.py::TestNN::test_ConvTranspose2d_groups_cuda_tf32 - RuntimeError: cuDNN error: CUDNN_STATUS_BAD_PARAM
  ```

Although this is not a complete implementation of cuDNN v8 API binding, I still want to merge this first. This would allow me to do small and incremental work, for the ease of development and review.

### CODE REVIEW

File: aten/src/ATen/native/cudnn/Macros.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <ATen/cudnn/cudnn-wrapper.h>
+
+// Warning: The following macro is guard for enabling cuDNN v8 API,
+// which is not finished yet on PyTorch. This macro will be removed
+// once the cuDNN v8 binding is done.
+// cuDNN v8 API is not finished yet, and is not recommended to use.
+// enable this only if you know what you are doing.
+#define _ENABLE_CUDNN_V8_API true
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cudnn/Macros.h:10]: I will change this to `false` before merge
```suggestion
#define _ENABLE_CUDNN_V8_API false
```

**Comment by zasdfgbnm** [comment]
Failures are because pytorch treats warnings as errors, and warnings in cudnn frontend causes compilation error. I will ask cudnn to fix these warnings.

**Comment by ngimel** [comment]
Can you assess the damage (if any) with instruction counts https://pytorch.org/docs/master/benchmark_utils.html#torch.utils.benchmark.Timer.collect_callgrind? So many std::vectors, so many std::copies. Our current bindings are not particularly efficient also, with TensorArgs, so maybe it nets out ok. 

cc @ezyang

**Comment by zasdfgbnm** [comment]
@ngimel OK, I will try with `collect_callgrind`. But this PR doesn't mean to be fast or feature complete. I haven't done any kind of benchmark here. This PR just build a sketch so that future work can be done incrementally based on this. So I hope benchmarks/perf improvements can be done in future PRs.

File: aten/src/ATen/native/cudnn/Conv_v8.cpp [hunk 2]
```diff
@@ -1,5 +1,177 @@
...
+#include <unordered_map>
+
+namespace at { namespace native{
+
+namespace {
+
+uint8_t getAlignment(const Tensor &t) {
+  // alignment are in bytes
+  uint8_t alignment = 1;
+  uint64_t address = reinterpret_cast<uint64_t>(t.data_ptr());
```
Comment (ezyang) [inline aten/src/ATen/native/cudnn/Conv_v8.cpp:29]: cast to `uintptr_t` is more standards complaint

**Comment by ezyang** [comment]
I too am curious how much slower this integration is, although more on the "everything gets slower when you port it to C++" front haha.

Well, if this is where NVIDIA is headed, we gotta head that way too. @ngimel given that this isn't on by default, I guess we can get started with this?

**Comment by ngimel** [comment]
Yeah, sure, we can get started with it. There are different ways of heading to C++, with lots of std::vectors, say, and without :-). But cudnn-frontend is open source and not set in stone, so if it is a problem, we can work with them to make it faster.

**Review by ngimel — APPROVED** [review]

**Comment by zasdfgbnm** [comment]
cc @ngimel @ezyang 

The callgrind result for the current v7 and v8 API are:

v7 API
```
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7f424e1277c0>
conv(input)
setup: import torch; conv = torch.nn.Conv2d(1, 1, 1).cuda(); input = torch.randn((100, 1, 100, 100), device="cuda")
                           All          Noisy symbols removed
    Instructions:     12473820                   12473820
    Baseline:              895                        895
100 runs per measurement, 1 thread
```

v8 API
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7f16b492d580>
conv(input)
setup: import torch; conv = torch.nn.Conv2d(1, 1, 1).cuda(); input = torch.randn((100, 1, 100, 100), device="cuda")
  51.95 us
  1 measurement, 1000000 runs , 1 thread
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7f16b492d220>
conv(input)
setup: import torch; conv = torch.nn.Conv2d(1, 1, 1).cuda(); input = torch.randn((100, 1, 100, 100), device="cuda")
                           All          Noisy symbols removed
    Instructions:     13585372                   13585372
    Baseline:              897                        897
100 runs per measurement, 1 thread
```
code:
```python
import torch
import torch.utils.benchmark

conv = torch.nn.Conv2d(1, 1, 1).cuda()
input = torch.randn((100, 1, 100, 100), device='cuda')

t = torch.utils.benchmark.Timer(
    stmt='conv(input)',
    setup='import torch; conv = torch.nn.Conv2d(1, 1, 1).cuda(); input = torch.randn((100, 1, 100, 100), device="cuda")')

result = t.timeit()
print(result)

result = t.collect_callgrind()
print(result)
```

**Comment by ngimel** [comment]
Both those numbers are so ridiculously bad :-)

**Pull request closed without merging**